### ISSUE TITLE
DeepExplainer+softmax no correlation between input and output (with reproducible code)

### ISSUE THREAD

**Issue by ydib** [shap/shap#1506]
TLDR: I couldn't make DeepExplainer show the correlation between input and output when using a softmax, the plots are below, the code is [here](https://github.com/ydib/shap_softmax_problem/blob/main/softmax_problem.py).

### Introduction

First of all, well done and thank you for shap. I have been using it extensively for my master thesis about feature explanation in AI for medicine.

I have come across an issue I would like to share with you, maybe you have some insight about what the problem is. You can find the code which generates all the plots below [here](https://github.com/ydib/shap_softmax_problem/blob/main/softmax_problem.py).

The packages used are:
> `numpy==1.18.5`
> `matplotlib==3.2.1`
> `tensorflow==2.2.0`
> `shap==0.37.0`

### Experiment

The goal in this toy example is to use the different explainers provided by shap to compute the shapley values for a model fitted to a very simple dataset. The model takes as input two real numbers (x0,x1) and outputs (y0,y1).
- (x0,x1) are generated using a normal distribution (mean = 100, std = 50).
- y0=1 if x0>=100, and 0 otherwise.
- y1=1 if x0<100, and 0 otherwise. Hence y1=-y0.
- x1 doesn't influence y0 and y1.
So y0(x0) is a step function centered at 100.

To fit this dataset I use two simple models with the same architecture:
```
_________________________________________________________________
Layer (type)                 Output Shape              Param #   
=================================================================
dense (Dense)                (None, 32)                96        
_________________________________________________________________
dense_1 (Dense)              (None, 16)                528       
_________________________________________________________________
dense_2 (Dense)              (None, 2)                 34        
=================================================================
Total params: 658
Trainable params: 658
Non-trainable params: 0
_________________________________________________________________
```
One of them has a sigmoid activation function at the dense_2 layer, and the other one a softmax. After training, both models have an accuracy of more than 0.9.

I then use DeepExplainer, GradientExplainer and KernelExplainer to compute the shap values, and use the summary_plot for the y0 output. Below are the 6 plots.

### Discussion
What we can observe is for all combinations of explainer/activation function there is a clear correlation between x0 and y0, except for the DeepExplainer with the softmax. Even after reading the papers cited in the project, I couldn't explain why that was the case.

### Question
Does anyone have an explanation for why DeepExplainer doesn't show a correlation between x0 and y0 when using a softmax ?

![DeepExplainer, sigmoid, y0](https://user-images.githubusercontent.com/15943664/95344230-b72bb900-08b9-11eb-8132-d1d320b16575.png)
![DeepExplainer, softmax, y0](https://user-images.githubusercontent.com/15943664/95344256-be52c700-08b9-11eb-85af-ee0adfdc8cfd.png)
![GradientExplainer, sigmoid, y0](https://user-images.githubusercontent.com/15943664/95344262-c01c8a80-08b9-11eb-8ae4-7effaf1ae329.png)
![GradientExplainer, softmax, y0](https://user-images.githubusercontent.com/15943664/95344269-c1e64e00-08b9-11eb-8931-3cd2fd10041d.png)
![KernelExplainer, sigmoid, y0](https://user-images.githubusercontent.com/15943664/95344272-c3177b00-08b9-11eb-988c-e3798638e993.png)
![KernelExplainer, softmax, y0](https://user-images.githubusercontent.com/15943664/95344281-c448a800-08b9-11eb-8651-b0a1a858e182.png)

**Comment by rabyj** [comment]
I also had a similar problem with the SHAP values not making sense for a dense feedforward neural network (couldn't match the model output by using expected value + sum of SHAP values, which is supposed to always be true), and had to modify my model to output logits instead of directly probability scores. Then it summed correctly. This did make the interpretation of the values themselves way more difficult though.

@slundberg did you ever encounter a similar problem?

**Comment by NightMachinery** [comment]
Are there any equations/papers for how Shapley is computed for softmax? They surely don't use the combinatoric formula, as that is too expensive.

**Comment by rabyj** [comment]
@NightMachinery 

> Are there any equations/papers for how Shapley is computed for softmax? They surely don't use the combinatoric formula, as that is too expensive.

I don't think whether softmax is the output of the model or not changes how the SHAP values are computed. The specific way the shap values are computed are different for each method.

The doc [here](https://shap-lrjball.readthedocs.io/en/latest/generated/shap.DeepExplainer.html) specifies how DeepExplainer SHAP values are computed.

> This is an enhanced version of the DeepLIFT algorithm (Deep SHAP) where, similar to Kernel SHAP, we approximate the conditional expectations of SHAP values using a selection of background samples. Lundberg and Lee, NIPS 2017 showed that the per node attribution rules in DeepLIFT (Shrikumar, Greenside, and Kundaje, arXiv 2017) can be chosen to approximate Shapley values. By integrating over many backgound samples Deep estimates approximate SHAP values such that they sum up to the difference between the expected model output on the passed background samples and the current model output (f(x) - E[f(x)]).

Paper is [A Unified Approach to Interpreting Model
Predictions](https://proceedings.neurips.cc/paper_files/paper/2017/file/8a20a8621978632d76c43dfd28b67767-Paper.pdf) (2017)

**Comment by NightMachinery** [comment]
@rabyj Thanks. Are there any papers about closed forms of the exact Shapley value, or at least its exact approximation (as opposed to a probabilistic approach that contrasts with "background" samples)? Softmax has only O(tokens) inputs, and an exact solution might be possible.

**Comment by rabyj** [comment]
@NightMachinery  Sorry, I'm unsure what you mean, I don't know. The paper I shared gives the exact formula used for DeepExplainer as far as I know. (**Deep SHAP (DeepLIFT + Shapley values)** section).

Going to tag some active committers since no one from the more active team seems to have ever taken a look at this issue, which seems to be a bit of a major problem with DeepExplainer.

@CloseChoice @connortann. Sorry if you're super busy guys but this seems kind of important. Would appreciate if anyone could assess this issue, OP is pretty good.

**Comment by CloseChoice** [comment]
Thanks for bringing this back up. I would agree that this is kind of important but not sure how much time I have to dig into this since a fix is still some 10-20 hours away.

So [here](https://github.com/shap/shap/blob/master/shap/explainers/_deep/deep_tf.py#L464) is how shap currently handles softmax. I haven't looked into this thoroughly but have a rough understanding what should go on theoretically but would need to work out the exact formula. 

I think if a fix is important to you then there is a lot we can do before actually going to the implementation. Here would be a rough plan:
 - [ ] rewrite the reproducible example in pytorch and use [captum's](https://github.com/pytorch/captum) DeepLiftShap method and verify that the result isn't different there
 - [ ] if the first point shows that something is off (which I believe given the information) we would need to figure out the concrete formula we need to apply in the softmax case
 - [ ] in parallel to case 2 we could write a test specifically for the softmax function referred above. So we would like to be able to test only this function and check it's return values to make this issue as easily reproducible as possible.

I can take over point 2 once the first is verified.

**Comment by rabyj** [comment]
Thanks so much @CloseChoice !

@ydib still alive and kicking? Want to take care of this? This wouldn't be my priority since I have a paper deadline breathing down my neck.

**Comment by NightMachinery** [comment]
Thanks. I meant the original Shapley formula as given in Wikipedia, which is not probabilistic at all:
![image](https://github.com/shap/shap/assets/36224762/34f0f3db-162b-4009-ad75-985f42bc1a76)
